rclrs/src/parameter/service.rs

```diff
@@ -1,89 +1,264 @@
+                        let name = name.to_cstr().to_str().ok()?;
+                        let storage = map.storage.get(name)?;
+                        let mut descriptor = match storage {
+                            ParameterStorage::Declared(storage) => {
+                                let (integer_range, floating_point_range) =
+                                    storage.options.ranges.to_descriptor_ranges();
+                                ParameterDescriptor {
+                                    name: name.into(),
+                                    type_: Default::default(),
+                                    description: storage.options.description.clone().into(),
+                                    additional_constraints: storage
+                                        .options
+                                        .constraints
+                                        .clone()
+                                        .into(),
+                                    dynamic_typing: matches!(storage.kind, ParameterKind::Dynamic),
+                                    read_only: matches!(storage.value, DeclaredValue::ReadOnly(_)),
+                                    floating_point_range,
+                                    integer_range,
+                                }
+                            }
+                            ParameterStorage::Undeclared(_) => ParameterDescriptor {
+                                name: name.into(),
+                                dynamic_typing: true,
+                                ..Default::default()
+                            },
+                        };
+                        descriptor.type_ = storage.to_parameter_type();
+                        Some(descriptor)
+                    })
+                    .collect::<Option<_>>()
+                    .unwrap_or_default();
+                // TODO(luca) error logging if the service call failed
+                DescribeParameters_Response { descriptors }
             },
         )?;
-        let get_parameter_types_service = Service::new(Arc::clone(&rcl_node_mtx),
-            "get_parameter_types",
-            |req_id: &rmw_request_id_t, req: GetParameterTypes_Request| {
-                GetParameterTypes_Response {
-                    types: seq![]
-                }
+        let map = parameter_map.clone();
+        let get_parameter_types_service = node.create_service(
+            &(fqn.clone() + "/get_parameter_types"),
+            move |_req_id: &rmw_request_id_t, req: GetParameterTypes_Request| {
+                let map = map.lock().unwrap();
+                let types = req
+                    .names
+                    .into_iter()
+                    .map(|name| {
+                        let name = name.to_cstr().to_str().ok()?;
+                        map.storage.get(name).map(|s| s.to_parameter_type())
+                    })
+                    .collect::<Option<_>>()
```

Comment:
If I'm grokking this correctly, if the user included even **one** parameter name in the request that doesn't exist in the storage, we're going to end up collecting the **whole** sequence (including any valid parameter names) into a single `None` and then `unwrap_or_default()` below will produce an empty sequence.

According to [the documentation for GetParameterTypes](https://docs.ros2.org/galactic/api/rcl_interfaces/srv/GetParameterTypes.html) that would not be the correct behavior. Instead we should return a value of `PARAMETER_NOT_SET` in the same location in the sequence where the unknown parameter was placed. I think we can fix this by simply replacing this and the next line with
```
.map(|type_| type_.unwrap_or(ParameterType::PARAMETER_NOT_SET)
.collect();
```

Reply:
I guess this is one of the cases where documentation is different from actual behavior.
I agree that it is puzzling but all of these were actually a conscious choice. What I did was try these edge cases on `rclcpp` and check how it behaves. Since this is for `get_parameter_types` here is a snippet from my local testing under iron:

```
lucadv@ros:~$ ros2 service call /minimal_publisher/get_parameter_types rcl_interfaces/srv/GetParameterTypes "{names: [use_sim_time]}"       
requester: making request: rcl_interfaces.srv.GetParameterTypes_Request(names=['use_sim_time'])

response:
rcl_interfaces.srv.GetParameterTypes_Response(types=[1])

lucadv@ros:~$ ros2 service call /minimal_publisher/get_parameter_types rcl_interfaces/srv/GetParameterTypes "{names: [use_sim_time, hello]}"
requester: making request: rcl_interfaces.srv.GetParameterTypes_Request(names=['use_sim_time', 'hello'])

response:
rcl_interfaces.srv.GetParameterTypes_Response(types=[])
```

As you can see, the actual behavior of `rclcpp` is that it returns an empty array as soon as one of the names is invalid.
The implementation for get_parameter_types is [here](https://github.com/ros2/rclcpp/blob/1c350d0d7fb9c7158e0a39057112486ddbd38e9a/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L1007-L1030), you can see how in that implementation the message creation actually throws an exception if any parameter does not exist, and that exception will be caught downstream and send an empty response.
This same behavior is present in `rclpy`, [here](https://github.com/ros2/rclpy/blob/7a7f23e0d7e51dd244001ef606b97f1153e5a97e/rclpy/rclpy/node.py#L615-L646).
What I did _not_ notice however, until you pointed it out and I went to cross check, is that the behavior you mention is indeed what happens when nodes allow undeclared parameters, so I will fix this now and add a test case for it.

Reply:
Interesting... Iirc restricting to only declared parameters was a later add on of ROS 2, not in the initial design. I suspect when ability to restrict it was added, the person implementing it was not mindful of the documented behavior of the service. This is probably something to report to the other client library maintainers.

Reply:
For now I just addressed the difference in behavior between my implementation and the other client libraries in https://github.com/ros2-rust/ros2_rust/pull/342/commits/136047989bd259c46a5638ff41a91430b8f5835a.
Depending on how the discussion on other client libraries goes I'm happy to revert it but at least for now we behave the same way.
For what it's worth I backtracked to this PR https://github.com/ros2/rclcpp/pull/495 when the change was introduced. Before the change the behavior was what you mentioned, after the change it is the current one.

Reply:
Changed the behavior and tests to return `PARAMETER_NOT_SET`. https://github.com/ros2-rust/ros2_rust/pull/342/commits/3274f42ce33ae094bd4b2020a03f72dec75d716b.
The same applies for all other services but I'll just bring it up here that the closure still returns an `Option` and the `unwrap_or_default` that results in an empty response is still present.
The failure scenario where an empty array will be returned is when conversion to string for the requested parameter name failed. I admit not really knowing how to handle this, since we can't convert the parameter to a string we also can't safely return a `PARAMETER_NOT_SET` response for it since that requires... returning its name in a string format.
I'm happy to change this if any other suggestion comes up.